### PR TITLE
Bug fix loss func

### PR DESCRIPTION
**Pull request by b-biswas**
previously the function was returning a loss of dimension [batch_size, 59, 59, 6], where 59, 59, 6 is the dimension of input image. 
This is not the expected dimension for a custom TensorFlow loss function. 

Now with a scalar return type, the model trains much better.

### CODE REVIEW

**Review by aboucaud — COMMENTED** [review]
I'm a bit confused.
It's not really a bug, in the sense that the loss as it was defined probably worked until some point.
At least I've seen it defined that way for a long time in tfp.

But the way things are computed may have changed. Anyway my point is I don't understand why there should be both a sum followed by an average on the batch dim.

After looking at https://github.com/tensorflow/probability/blob/main/tensorflow_probability/examples/vae.py I see that they do use a way to reduce to a scalar at the end, but they only use the `tf.reduce_mean()` on the result NLL of the decoder. There is no sum used here.

**Comment by aboucaud** [comment]
Also here an other exemple where this is not needed.
https://www.tensorflow.org/probability/examples/Probabilistic_Layers_VAE

I think what I don't understand is why you end up having the full tensor shape as output of the `log_prob`. There is probably a confusion between event shape and sample shape as explained in 
https://www.tensorflow.org/probability/examples/Understanding_TensorFlow_Distributions_Shapes

Note that I am saying all these things without touching the code (don't have time right now) so I might completely by wrong.

@BastienArcelin want to comment ?

**Comment by b-biswas** [comment]
It's ok to do a sum or a mean. 
Ultimately what changes is a constant for multiplication. 

It's ok to do just a tf.math.mean() on the entire array, it's just that the coefficient of the kl divergence and probably the lr that needs to be updated accordingly. 

You are right about the first example. It will work because the return type still has a dimension of [batch_size].
When you return something of the dimension [batch_size] TensorFlow implicitly performs the mean. (I have already checked that these are giving the exact same values for loss, but Junpeng told me to take the mean explicitly anyway so I did). 

The problem arises when there are more dimensions to the loss. 

I don't know the exact specifics here but If you refer to [https://www.tensorflow.org/api_docs/python/tf/keras/losses/Loss](https://www.tensorflow.org/api_docs/python/tf/keras/losses/Loss)
the return type has a different dimension from the input. 
To avoid confusion I think it is better to stick to the scalar. It definitely works always.

**Comment by b-biswas** [comment]
> Also here an other exemple where this is not needed. https://www.tensorflow.org/probability/examples/Probabilistic_Layers_VAE
> 
> I think what I don't understand is why you end up having the full tensor shape as output of the `log_prob`. There is probably a confusion between event shape and sample shape as explained in https://www.tensorflow.org/probability/examples/Understanding_TensorFlow_Distributions_Shapes
> 
> Note that I am saying all these things without touching the code (don't have time right now) so I might completely by wrong.
> 
> @BastienArcelin want to comment ?

The prediction is a distribution on each pixel and log prob gives pixel-wise log_prob here to the return type should have image dimension but I did not check it.

File: debvader/metrics.py [hunk 1]
```diff
@@ -17,10 +18,14 @@ def vae_loss(ground_truth, predicted_distribution):
     """
     computes the reconstaruction loss term for the VAE.
```
Comment (aboucaud) [inline debvader/metrics.py:19]:
```suggestion
    computes the reconstruction loss term for the VAE.
```

File: debvader/metrics.py [hunk 2]
```diff
@@ -17,10 +18,14 @@ def vae_loss(ground_truth, predicted_distribution):
...
 
-    returns the log_prob of the ground truth under the predicted distribution
+    returns the negative log_prob of the ground truth under the predicted distribution
 
     parameters:
         ground_truth: original ground truth image
         predicted_distribution: distribution predicted by network
```
Comment (aboucaud) [inline debvader/metrics.py:25]:
```suggestion
        predicted_distribution: output distribution from the decoder network
```

**Comment by aboucaud** [comment]
> The prediction is a distribution on each pixel and log prob gives pixel-wise log_prob here to the return type should have image dimension but I did not check it.

Right I get it now, the sum turns the pixel-wise log_prob into a global one for the image and the mean averages over the batch. Makes sense. Then an explanation in the function could be nice, and maybe `scalar` added to the function name.

**Comment by b-biswas** [comment]
@aboucaud I added a description in the function docstring. Maybe we can merge this for now?

**Review by aboucaud — APPROVED** [review]
